crates/fuel-core/src/graphql_api/da_compression.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    fuel_core_graphql_api::ports::worker::OffChainDatabaseTransaction,
+    graphql_api::storage::da_compression::{
+        DaCompressedBlocks,
+        DaCompressionTemporalRegistry,
+        DaCompressionTemporalRegistryEvictor,
+        DaCompressionTemporalRegistryIndex,
+    },
+};
+use fuel_core_compression::{
+    compress::compress,
+    ports::{
+        EvictorDb,
+        TemporalRegistry,
+        UtxoIdToPointer,
+    },
+    RegistryKeyspace,
+    RegistryKeyspaceValue,
+};
+use fuel_core_storage::{
+    not_found,
+    StorageAsMut,
+    StorageAsRef,
+};
+use fuel_core_types::{
+    blockchain::block::Block,
+    fuel_tx::{
+        input::PredicateCode,
+        Address,
+        AssetId,
+        ContractId,
+        ScriptCode,
+    },
+    services::executor::Event,
+};
+use futures::FutureExt;
+
+/// Performs DA compression for a block and stores it in the database.
+pub fn da_compress_block<T>(
+    block: &Block,
+    block_events: &[Event],
+    db_tx: &mut T,
+) -> anyhow::Result<()>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let compressed = compress(
+        CompressTx {
+            db_tx,
+            block_events,
+        },
+        block,
+    )
+    .now_or_never()
+    .expect("The current implementation resolved all futures instantly")?;
+
+    db_tx
+        .storage_as_mut::<DaCompressedBlocks>()
+        .insert(&block.header().consensus().height, &compressed)?;
+
+    Ok(())
+}
+
+struct CompressTx<'a, Tx> {
+    db_tx: &'a mut Tx,
+    block_events: &'a [Event],
+}
+
+macro_rules! impl_temporal_registry {
+    ($($name:ident: $type:ty),*$(,)?) => {
+        $(
+            impl<'a, Tx> TemporalRegistry<$type> for CompressTx<'a, Tx>
+            where
+                Tx: OffChainDatabaseTransaction,
+            {
+                fn read_registry(
+                    &self,
+                    key: fuel_core_types::fuel_compression::RegistryKey,
+                ) -> anyhow::Result<$type> {
+                    let v = self
+                        .db_tx
+                        .storage_as_ref::<DaCompressionTemporalRegistry>()
+                        .get(&(RegistryKeyspace::$name, key))?
+                        .ok_or(not_found!(DaCompressionTemporalRegistry))?
+                        .into_owned();
+                    match v {
+                        RegistryKeyspaceValue::$name(v) => Ok(v),
+                        _ => anyhow::bail!("Unexpected value in the registry"),
+                    }
+                }
+
+                fn write_registry(
+                    &mut self,
+                    key: fuel_core_types::fuel_compression::RegistryKey,
+                    value: $type,
+                ) -> anyhow::Result<()> {
+                    // Write the actual value
+                    self.db_tx
```

Comment:
I think we do too many clones=) While it is okay for `ContractId` and similar types, it can be an issue for script and predicate. If we used separate tables for each type we could avoid it

Reply:
Clones are removed in 946a2e8f23fef2109fb9ba286ebd2d451c2af7cb